File: src/ann_index.rs
use hnsw_rs::prelude::*;

pub type ANNResult<T> = Result<T, Box<dyn std::error::Error>>;

pub trait ANNIndex<'a> {
    fn init_index(&mut self, max_elements: usize, dimension: usize) -> ANNResult<()>;
    fn add_pt(&mut self, point: Vec<f32>, id: usize) -> ANNResult<()>;
    fn get_curr_ct(&self) -> ANNResult<(usize)>;
    fn get_max_elements(&mut self) -> ANNResult<(usize)>;
    fn resize(&mut self, new_size: usize) -> ANNResult<()>; 
    fn search_knn(&self, query: &[f32], k: usize) -> ANNResult<Vec<(usize, f32)>>;
}


pub struct HnswAnnIndex<'a> {
    hnsw: Hnsw<'a, f32, DistCosine>,
    dimension: usize,
    max_elements: usize,
    points: Vec<(Vec<f32>, usize)>,
}

impl<'a> ANNIndex<'a> for HnswAnnIndex<'a> {
    fn init_index(&mut self, max_elements: usize, dimension: usize) -> ANNResult<()> {
        self.hnsw = Hnsw::new(16, max_elements, dimension, 16, DistCosine);
        self.dimension = dimension;
        self.max_elements = max_elements;
        self.points = Vec::new();
        return Ok(());
    }

    fn add_pt(&mut self, point: Vec<f32>, id: usize) -> ANNResult<()> {
        if point.len() != self.dimension {
            return Err("Point dimensions don't match!".into());
        }

        self.hnsw.insert((&point, id));
        self.points.push((point, id));
        
        return Ok(());
    }

    fn get_curr_ct(&self) -> ANNResult<(usize)> {
        let ct = self.hnsw.get_nb_point();
        return Ok(ct);
    }

    fn get_max_elements(&mut self) -> ANNResult<(usize)> {
        return Ok(self.max_elements);
    }

    fn resize(&mut self, new_size: usize) -> ANNResult<()> {
        let hnsw = Hnsw::new(16, new_size, self.dimension, 16, DistCosine);
        
        // Reinsert each stored point into the new index.
        for (point, id) in &self.points {
            hnsw.insert((&point, *id));
        }

        self.hnsw = hnsw;
        self.max_elements = new_size;

        return Ok(());
    }

    fn search_knn(&self, query: &[f32], k: usize) -> ANNResult<Vec<(usize, f32)>> {
        let res = self.hnsw.search(query, k, k+2);

        let knn_results: Vec<(usize, f32)> = res.into_iter()
        .map(|result| (result.d_id, result.distance))
        .collect();

        return Ok(knn_results);
    }

    
}

File: src/ann_index.py
from abc import ABC, abstractmethod
import hnswlib
from typing import List, Tuple


class ANNIndex(ABC):
    @abstractmethod
    def init_index(self, max_elements: int, dimension: int) -> None:
        ...

    @abstractmethod
    def add_pt(self, point: List[float], id: int) -> None:
        ...

    @abstractmethod
    def get_curr_ct(self) -> int:
        ...

    @abstractmethod
    def get_max_elements(self) -> int:
        ...

    @abstractmethod
    def resize(self, new_size: int) -> None:
        ...

    @abstractmethod
    def search_knn(self, query: List[float], k: int) -> List[Tuple[int, float]]:
        ...


class HnswAnnIndex(ANNIndex):
    def __init__(self, max_elements: int, dimension: int):
        self.dimension = dimension
        self.max_elements = max_elements
        self.points: List[Tuple[List[float], int]] = []
        self._init_hnsw()

    def _init_hnsw(self):
        self.index = hnswlib.Index(space='cosine', dim=self.dimension)
        # M=16, ef_construction=100 are defaults you can tune
        self.index.init_index(max_elements=self.max_elements, ef_construction=100, M=16)

    def init_index(self, max_elements: int, dimension: int) -> None:
        self.dimension = dimension
        self.max_elements = max_elements
        self.points.clear()
        self._init_hnsw()

    def add_pt(self, point: List[float], id: int) -> None:
        if len(point) != self.dimension:
            raise ValueError("Point dimensions don't match!")
        self.index.add_items(point, ids=[id])
        self.points.append((point, id))

    def get_curr_ct(self) -> int:
        return self.index.get_current_count()

    def get_max_elements(self) -> int:
        return self.max_elements

    def resize(self, new_size: int) -> None:
        # Rebuild a larger index
        old_points = list(self.points)
        self.max_elements = new_size
        self._init_hnsw()
        for pt, pid in old_points:
            self.index.add_items(pt, ids=[pid])
        self.points = old_points

    def search_knn(self, query: List[float], k: int) -> List[Tuple[int, float]]:
        labels, distances = self.index.knn_query(query, k=k)
        # hnswlib returns lists of arrays
        return list(zip(labels[0].tolist(), distances[0].tolist()))
